### ISSUE TITLE
Replicating an envvar in a modextravars entry?

### ISSUE THREAD

**Issue by kelseymh** [easybuilders/easybuild-easyconfigs#11049]
I have a (local) EasyConfig for some of our group software which depends upon Matlab.  Naturally, the group software doesn't know about EasyBuild, so it has its own expected envvar for the Matlab installation directory.  In my EasyConfig, I want to set this group envvar to match the default `$MATLABROOT` value which gets set by `module load Matlab`.

What I wrote, taking my cue from the obvious shell version, was:

    modextravars = {
        'MATLABINSTALL': '$MATLABROOT',
        'MLM_LICENSE_FILE': '/sw/hprc/sw/Matlab/licenses/license.dat',
    }

but this fails.  The .lua file ends up with a quoted string:

    setenv("MATLABINSTALL", "$MATLABROOT")

and the shell ends up with an unexpanded, literal expression:

    [kelsey@terra3 eb]$ env|grep MATLAB
    MATLABINSTALL=$MATLABROOT
    MATLABDIR=/sw/hprc/sw/Matlab
    MATLABROOT=/sw/hprc/sw/Matlab/R2019b

etc.  Is there an easy way ask to have an envvar expanded like this in `modextravars`?  Failing that, is there a specific way to capture one of the `setenv()` settings from a different module (in this case Matlab)?

**Comment by casparvl** [comment]
Out of the top of my head (and I haven't tested this) the easiest is probably to use the `%(installdir)s` template instead of the `$MATLABROOT`. I.e.
```
modextravars = {
    'MATLABINSTALL': %(installdir)s,
    'MLM_LICENSE_FILE': '/sw/hprc/sw/Matlab/licenses/license.dat',
}
```

Since modextravars is all about generating the modulefile, `MATLABROOT` would not be set yet at that time anyway, so there is no point in trying to read that environment variable at that time. 

If you ever REALLY need to pick up on an environment variable at the time the module is loaded, e.g. because you want to set `MYSPECIFICVAR` equal to `TMPDIR` at the time the module is loaded, you can use modtclfooter or modluafooter to add some real `tcl` or `lua` code in your easyconfig. E.g. for TCL:
```
modtclfooter = """setenv MYSPECIFICVAR $env(TMPDIR)
```
However, this makes EasyConfigs less generic and generally more tuned to either work with LUA ór TCL modules, so if it can be avoided, that's better.

**Comment by kelseymh** [comment]
Thanks, @casparvl !  The `%(installdir)s` option isn't appropriate for my case.  That would work if we were doing a Matlab module ourselves.  On our cluster, Matlab is installed as an `EXTERNAL_MODULE` with a hand-written module (.lua) file.  So I'm trying to get access to one of the envvars which that other module sets.

I had thought about your idea for using `modtclfooter`, but I'm of the same opinion as you; it's not generic.  I guess that I could make a custom EasyBlock for this, and use the Lmod Python functionality to generate an output line with the getenv functionality.

**Comment by casparvl** [comment]
Ah, sorry, I misunderstood: this is an EasyConfig for your 'other' (not Matlab) software.

So then, it ís loaded as dep, and the env var should be set. Hm, I'm not sure if there is an (elegant) way to get an environment variable. Of course, EasyConfigs are 'just' Python, but the traditional way of getting an OS env var through `os.getenv` won't work: this would be executed when the EasyConfig is parsed, and at thát time, the environment variable is nót set yet.

You might be able to use a hook to do this: https://easybuild.readthedocs.io/en/latest/Hooks.html I'm guessing with the pre-module hook, you could use os.getenv to grep the environment variable, and add it to the modextravars... Not sure if this is _the_ most elegant way to do it, but I expect it would work.

I know that I myself have also fought with this at some point... Let me see if I can find that issue, it may prove useful...

**Comment by kelseymh** [comment]
> I know that I myself have also fought with this at some point...

I was pretty sure I wasn't the first :-)  Yeah, a hook might work for this, or an EasyBlock with `make_module_extra` overloaded.  Either way, I can then call `os.getenv` and plug in the value from the dependency load.

I've run into this sort of "customization" issue with a few of our group-software (i.e., not central installations) modules.  It seems like package-specific EasyBlocks are the preferred approach, rather than special code in the EasyConfig (which then has to be properly copied from version to version).

**Comment by casparvl** [comment]
Found it, I posted it on the EasyBuild user mailing list. It was somewhat similar in that I needed to have an environment variable point to the binary of another installation. I.e. module `foo` needed to set an environment variable `bar_bin` to point to `$EBROOTBAR/bin`. Suggestions solutions there were:
- simply have the `bar` module set `bar_bin`. This means `bar_bin` is always set when `bar` is loaded, even if only module `foo` needs it. That's not super elegant, but it is easy to implement, and would work.
- use a postinstallcmd to generate a symlink to `$EBROOTBAR/bin` in the `foo` installation directory:
```
postinstallcmds = ["ln -s $EBROOTBAR/bin/bar %(installdir)s/bin/bar"]
modextrapaths = {'BAR_BIN': '%(installdir)s/bin'}
```
The advantage of this is that the 'mess' is in the `foo` installation that requires it. However, your case isn't really equivalent and creating a symlink to the matlab root, just for the sake of having your `MATLABINSTALL` point to your symlink feels... more dirty.

Anyway, I think none of these three solutions (hooks, having the matlab module set it, or the symlink approach) are very elegant (though out of these three, I would probably prefer the hooks). I agree with you that a package-specific EasyBlock would be the most elegant & "EasyBuild-like" solution (at least imho).
